### PR TITLE
Rename getEntry to getEntryBySlug

### PR DESCRIPTION
**Pull request by matthewp**
## Changes

- Rename `getEntry` to `getEntryBySlug`.
  - The new function takes a slug as the second argument instead of an id.
  - It loops over the collection and finds a matching entry that has the provided slug.
- The types here could maybe be better? I could figure out a way to type it so that the provided slug matches an entry with that slug. I think the fact that a slug can be any `string` is a blocker here. Happy to be proven otherwise.

## Testing

- Tests updated

## Docs

- PR https://github.com/withastro/docs/pull/2408

### CODE REVIEW

**Review by bholmesdev — COMMENTED** [review]
Just some type cleanup. Some `entry`guing changes!

File: packages/astro/src/content/template/types.d.ts [hunk 1]
```diff
@@ -30,17 +30,21 @@ declare module 'astro:content' {
...
+	type EntryMapKeys = keyof typeof entryMap;
+	type AllValuesOf<T> = T extends any ? T[keyof T] : never;
+	type ValidEntrySlug<C extends EntryMapKeys> = AllValuesOf<typeof entryMap[C]>['slug'];
+
+	export function getEntryBySlug<C extends keyof typeof entryMap>(
 		collection: C,
-		entryKey: E
-	): Promise<(typeof entryMap)[C][E] & Render>;
+		// Note that this has to accept a regular string too, for SSR
+		entrySlug: ValidEntrySlug<C> | string
```
[the remark this answers is not in the document]
Comment (matthewp) [reply packages/astro/src/content/template/types.d.ts:40]: Ah ok, thank you. I realized that the SSR use-case you're going to be grabbing a string from Astro.params so not allowing any strings would give bad type messages. I'll try your suggestion.

[the remark this answers is not in the document]
Comment (matthewp) [reply packages/astro/src/content/template/types.d.ts]: Ok, updated here: https://github.com/withastro/astro/pull/5893/commits/d41cc2f6eacde017e5d83b7cae3f9a3ff3e56bd7

This works in my testing. If you provide a known slug value you get back just an entry. If you provide any other string you get back an entry or undefined, which is what we want.

File: packages/astro/src/content/internal.ts [hunk 2]
```diff
@@ -69,18 +69,28 @@ export function createGetCollection({
...
 }: {
-	collectionToEntryMap: CollectionToEntryMap;
+	getCollection: ReturnType<typeof createGetCollection>;
 	collectionToRenderEntryMap: CollectionToEntryMap;
 }) {
-	return async function getEntry(collection: string, entryId: string) {
-		const lazyImport = collectionToEntryMap[collection]?.[entryId];
-		if (!lazyImport) throw new Error(`Failed to import ${JSON.stringify(entryId)}.`);
+	return async function getEntryBySlug(collection: string, slug: string) {
+		const entries = await getCollection(collection);
```
Comment (FredKSchott) [inline packages/astro/src/content/internal.ts:80]: nit: might be good to add a comment/todo to improve to O(1) lookup in the future, just for any sleuths (or future contributors) who find this code and wonder why it was implemented this way.

**Review by FredKSchott — APPROVED** [review]
LGTM on the API! Thanks for getting this across the line in time for 2.0.

@bholmesdev I'm happy to factor these changes into my docs PR, so no action needed on your part other than giving a review if you can. cc @sarah11918 so she's aware!

**Pull request merged**